src/mam/mod_mam_muc_rdbms_arch.erl

```diff
@@ -109,541 +104,210 @@ stop_muc(Host) ->
-       "WHERE room_id = ", use_escaped_integer(escape_room_id(RoomID))]),
-    mongoose_rdbms:result_to_integer(BSize).
+-spec archive_size(Size :: integer(), Host :: jid:server(),
+                   ArcId :: mod_mam:archive_id(), ArcJID :: jid:jid()) -> integer().
+archive_size(Size, Host, ArcID, ArcJID) when is_integer(Size) ->
+    Filter = [{equal, room_id, ArcID}],
+    Env = env_vars(Host, ArcJID),
+    Result = lookup_query(count, Env, Filter, unordered, all),
+    mongoose_rdbms:selected_to_integer(Result).
+
+extend_params_with_sender_id(Host, Params = #{remote_jid := SenderJID}) ->
+    BareSenderJID = jid:to_bare(SenderJID),
+    SenderID = mod_mam:archive_id_int(Host, BareSenderJID),
+    Params#{sender_id => SenderID}.
 
 -spec archive_message(_Result, jid:server(), mod_mam:archive_message_params()) -> ok.
-archive_message(_Result, Host, Params = #{direction := incoming}) ->
+archive_message(_Result, Host, Params0 = #{local_jid := ArcJID}) ->
     try
-        Row = prepare_message(Host, Params),
-        {updated, 1} = mod_mam_utils:success_sql_execute(Host, insert_mam_muc_message, Row),
-        retract_message(Host, Params),
+        Params = extend_params_with_sender_id(Host, Params0),
+        Env = env_vars(Host, ArcJID),
+        do_archive_message(Host, Params, Env),
+        retract_message(Host, Params, Env),
         ok
-    catch _Type:Reason:StackTrace ->
-            ?LOG_ERROR(#{what => archive_message_failed,
-                         host => Host, mam_params => Params,
-                         reason => Reason, stacktrace => StackTrace}),
-            {error, Reason}
+    catch error:Reason:StackTrace ->
+              ?LOG_ERROR(#{what => archive_message_failed,
+                           host => Host, mam_params => Params0,
+                           reason => Reason, stacktrace => StackTrace}),
+              erlang:raise(error, Reason, StackTrace)
     end.
 
-retract_message(Host, #{archive_id := RoomID,
-                        remote_jid := SenderJID,
-                        packet := Packet}) ->
-    case mod_mam_utils:get_retract_id(mod_mam_muc, Host, Packet) of
+do_archive_message(Host, Params, Env) ->
+    Row = mam_encoder:encode_message(Params, Env, db_mappings()),
+    {updated, 1} = mod_mam_utils:success_sql_execute(Host, insert_mam_muc_message, Row).
+
+%% Retraction logic
+%% Called after inserting a new message
+-spec retract_message(jid:server(), mod_mam:archive_message_params()) -> ok.
+retract_message(Host, #{local_jid := ArcJID} = Params)  ->
+    Env = env_vars(Host, ArcJID),
+    retract_message(Host, Params, Env).
+
+-spec retract_message(jid:server(), mod_mam:archive_message_params(), env_vars()) -> ok.
+retract_message(Host, #{archive_id := ArcID, sender_id := SenderID,
+                        packet := Packet}, Env) ->
+    case get_retract_id(Packet, Env) of
         none -> ok;
-        OriginIDToRetract -> retract_message(Host, RoomID, SenderJID, OriginIDToRetract)
```

Comment:
FYI: the reason for the name `OriginIDToRetract` was not to confuse itwith the `OriginID` of the retraction message itself. Now it's a bit easier to confuse them.